File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.size = 0
        self.lt = []
        self.gt = []

    def addNum(self, num: int) -> None:
        self.size += 1
        
        if self.gt and num > self.gt[0]:
            heappush(self.gt, num)
            if len(self.gt) > self.size // 2:
                heappush(self.lt, -heappop(self.gt))
        else:
            heappush(self.lt, -num)
            if len(self.gt) < self.size // 2:
                heappush(self.gt, -heappop(self.lt))
        

    def findMedian(self) -> float:
        return -self.lt[0] if self.size & 1 else (-self.lt[0] + self.gt[0]) / 2
        


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
/*
left half will be maxheap
right half will be minheap
so that the top of both heaps will be the middle element(s)
*/
class MedianFinder {
public:
    MedianFinder() {
        length = 0;
    }
    
    void addNum(int num) {
        ++length;
        if (lefthalf.empty() || num <= lefthalf.top()) {
            lefthalf.push(num);
        } else {
            righthalf.push(num);
        }
        
        if (righthalf.size() > lefthalf.size()) {
            int rightmin = righthalf.top(); righthalf.pop();
            lefthalf.push(rightmin);
        } else if (lefthalf.size() - righthalf.size() == 2) {
            int leftmax = lefthalf.top(); lefthalf.pop();
            righthalf.push(leftmax);
        }
    }
    
    double findMedian() {
        if (length & 1) {
            return lefthalf.top();
        } else {
            return (lefthalf.top() + righthalf.top()) / 2.0;
        }
    }

private:
    int length;
    priority_queue<int> lefthalf;
    priority_queue<int, vector<int>, greater<int>> righthalf;
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */